### DISCUSSION TITLE
[Issue]: Can Autogen configure a network proxy

### DISCUSSION THREAD

**Discussion by qingpc** [microsoft/autogen#1854] [Q&A]
### Describe the issue

Can Autogen configure a network proxy to make calls to the official interfaces of AOAI or OpenAI through the network proxy.
Due to poor local network quality, I would like to set up a network proxy. Can it be done in Autogen? Where can I set it up?

### Steps to reproduce

The system proxy's https proxy and https proxy have been set, but they are not effective in the Autogen program. Do I need to directly add a proxy network proxy in the program

### Additional Information

autogenstudio  version:  0.0.47  
pyautogen version:  0.2.16

**Comment by ekzhu** [comment]
Have you checked out LLM proxies like Ollama and LiteLLM?

**Reply by qingpc** [reply]
We did not use local LLM or other models, but instead added the GPT4 key in autogenstudio-ui  and used the GPT4 API to obtain the results. Can I set up using a network proxy to call the API of GPT4?
Additionally, I will also consider how the local LLM model works，thank‘s

**Reply by jackgerrits** [reply]
It is possible to setup a network proxy for the OpenAI python client. See here: https://stackoverflow.com/questions/77606417/openai-api-request-with-proxy

However, the OpenAI client in autogen is a bit hidden. It may actually be easier to create a custom model client which uses OpenAI with your required network proxy configuration https://github.com/microsoft/autogen/blob/main/notebook/agentchat_custom_model.ipynb

**Reply by qingpc** [reply]
ok, maybe I need to Create and configure the custom model, let me try it, thank you very much for your help and suggestions, waiting for my good news

**Reply by qingpc** [reply]
The problem has been solved, thank you very much for your help!
Mainly by setting up the system's web proxy,

Option A:
``` 
 If using Docker containers:
Two operations are required:
1. Install the pip socket library for Docker images or containers
Pip install https [socks]
2. To start the Docker container, simply pass the proxy parameters:
For example:
-e http proxy=socks5://x.x.x.x:1080
-e https proxy=socks5://x.x.x.x:1080
```

Option B: 
```
If it is to directly execute the autostudio program
1. Install the pip socket library
Pip install https [socks]
2. Add system proxy environment variables, taking Linux as an example:
Export http proxy=socks5://x.x.x.x: 9999
Export https proxy=socks5://x.x.x.x: 9999
Export no_proxy='localhost, 127.0.0.1,. example. com '

Then execute the program. If using other process management tools, make sure to pass the environment variables to the starting process.
```

Then the program can use a network proxy to access GPT4 or OpenAI APIs

**Reply by sonichi** [reply]
Thanks. I remember this question was asked before. The solution was to add some field in the `llm_config`, if I remember it correctly.

**Reply by qingpc** [reply]
Ok, I will check the configuration of llm_config

**Reply by stnbSim** [reply]
hi @wiyisun, would you please share the solution on the llm_config's configuration?

**Reply by qingpc** [reply]
llm_config does not do special Settings, only need to set their own application key, I used GPT4 API KEY. You can configure this in the web interface provided by autogenstudio-ui, or in the autogen code root directory, where the configuration file name is required: OAI_CONFIG_LIST. Then launch autogenstudio-ui.

**Reply by ekzhu** [reply]
Documentation for how to do this: https://microsoft.github.io/autogen/docs/topics/llm_configuration#adding-http-client-in-llm_config-for-proxy

**Comment by ekzhu** [comment]
Yes. Please see the documentation here: https://microsoft.github.io/autogen/docs/topics/llm_configuration#adding-http-client-in-llm_config-for-proxy

**Comment by lc82111** [comment]
I have tested the solution in doc for Gemini, it's not working.

**Comment by prd-tu-tran** [comment]
Hey, I initiated an AgentBuilder class. It requires `config_file_or_env`. How can I save it to a file or set it up with a proxy? Do you guys know how to fix it?
